### PR TITLE
Rebase kubernetes to v1.18.0-beta.2

### PR DESCRIPTION
**Pull request by damemi**

### CODE REVIEW

**Comment by damemi** [comment]
This PR requires fixes to build, so /holding for now while I apply those

**Comment by damemi** [comment]
This doesn't compile because it's waiting on fixes to kube-storage-version-migrator:
```
    # github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:71:5: not enough arguments in call to c.client.Get().Resource("storagestates").Name(name).VersionedParams(&options, scheme.ParameterCodec).Do
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:87:5: not enough arguments in call to c.client.Get().Resource("storagestates").VersionedParams(&opts, scheme.ParameterCodec).Timeout(timeout).Do
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:103:8: not enough arguments in call to c.client.Get().Resource("storagestates").VersionedParams(&opts, scheme.ParameterCodec).Timeout(timeout).Watch
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:112:5: not enough arguments in call to c.client.Post().Resource("storagestates").Body(storageState).Do
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:124:5: not enough arguments in call to c.client.Put().Resource("storagestates").Name(storageState.ObjectMeta.Name).Body(storageState).Do
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:139:5: not enough arguments in call to c.client.Put().Resource("storagestates").Name(storageState.ObjectMeta.Name).SubResource("status").Body(storageState).Do
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:150:5: not enough arguments in call to c.client.Delete().Resource("storagestates").Name(name).Body(options).Do
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:165:5: not enough arguments in call to c.client.Delete().Resource("storagestates").VersionedParams(&listOptions, scheme.ParameterCodec).Timeout(timeout).Body(options).Do
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storagestate.go:177:5: not enough arguments in call to c.client.Patch(pt).Resource("storagestates").SubResource(subresources...).Name(name).Body(data).Do
            have ()
            want (context.Context)
    vendor/github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/clients/clientset/typed/migration/v1alpha1/storageversionmigration.go:71:5: not enough arguments in call to c.client.Get().Resource("storageversionmigrations").Name(name).VersionedParams(&options, scheme.ParameterCodec).Do
            have ()
            want (context.Context)
```

This is also currently in a bunch of commits that could be squashed down. Trying to switch to the `k8s.io/client-go/deprecated` package isn't quite as smooth as it seems. There's a lot of places that aren't using rest clients in that package, or that call out to external functions that expect the non-deprecated clients.

I also ran into a problem where I had to undo a lot of the deprecated client implementation (see https://github.com/openshift/library-go/pull/735/commits/2e4d36170763819c7a56364f29feb422340455e0), essentially because of [these 2 lines](https://github.com/openshift/library-go/blob/master/pkg/operator/staticpod/certsyncpod/certsync_cmd.go#L83-L85) where the informer expects a non-deprecated client, so if our events recorder takes a non-deprecated client that ends up cascading into a bunch of different places.

**Comment by damemi** [comment]
Updated to use `context` and client options (instead of switching to the deprecated client) and also to vendor the fork of kube-storage-version-migrator with fixes patched in. Builds locally so seeing now if CI goes green, and will finally squash to a sensible number of commits

**Comment by damemi** [comment]
/retest

```
=== RUN   TestSyncSecret
I0312 17:45:13.197669   10963 shared_informer.go:223] Waiting for caches to sync for ResourceSyncController
I0312 17:45:13.198047   10963 shared_informer.go:230] Caches are synced for ResourceSyncController 
I0312 17:45:13.198140   10963 base_controller.go:48] Starting #1 worker of ResourceSyncController controller ...
--- FAIL: TestSyncSecret (10.00s)
    resourcesync_controller_test.go:141: timeout while waiting for destination secret to be created
I0312 17:45:23.169862   10963 base_controller.go:95] Shutting down ResourceSyncController ...
```

Having trouble debugging this, because it passes locally:
```
$ go clean -testcache
$ go test -race -run TestSyncSecret ./pkg/operator/resourcesynccontroller/...
ok  	github.com/openshift/library-go/pkg/operator/resourcesynccontroller	1.584s
```

File: pkg/controller/metrics/client_metrics.go [hunk 1]
```diff
@@ -64,6 +64,10 @@ func (l *latencyAdapter) ClearState() {
 	l.m.Reset()
 }
 
+func (l *latencyAdapter) FQName() string {
+	return l.m.FQName()
```
Comment (damemi) [inline pkg/controller/metrics/client_metrics.go]: Not sure if these are the proper way to implement the `FQName()` function

File: pkg/image/dockerv1client/client.go [hunk 2]
```diff
@@ -34,46 +33,53 @@ var (
...
+	out.Parent = in.Parent
+	out.Comment = in.Comment
+	out.Created = in.Created.Time
+	out.Container = in.Container
+	out.DockerVersion = in.DockerVersion
+	out.Author = in.Author
+	out.Architecture = in.Architecture
+	out.Size = in.Size
+	return nil
+}
```
Comment (soltysh) [inline pkg/image/dockerv1client/client.go]: This is bad, the reason this was internal was to ensure nobody will use that, your approach suddenly makes those conversion functions public, at least make them private.

**Comment by damemi** [comment]
Added fix from https://github.com/openshift/library-go/pull/742

**Comment by ingvagabund** [comment]
One more change to carry: https://github.com/openshift/cluster-kube-controller-manager-operator/pull/376/commits/bd218906f66c484e7b1df29479a0274826bf82ef

**Comment by damemi** [comment]
> Another change to carry: openshift/cluster-kube-controller-manager-operator@bd21890
seems this is already in master so we're good here

File: pkg/controller/metrics/client_metrics.go (hunk 1 above)
Comment (damemi) [reply pkg/controller/metrics/client_metrics.go]: It seems like the update that brought this requirement (https://github.com/kubernetes/kubernetes/commit/4d028a7903ac66db5795056f654c8b0c03b7c90a) was minor and I think in this case using the metric's provided FQName() function is sufficient

File: pkg/image/dockerv1client/client.go (hunk 2 above)
Comment (damemi) [reply pkg/image/dockerv1client/client.go]: updated

**Pull request merged**